File: LeetCode/Medium/2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.java
class SmallestInfiniteSet {

    private final boolean[] arr;
    private int smallIdx;

    public SmallestInfiniteSet() {
        this.arr = new boolean[1_001];
        this.smallIdx = 1;
    }
    
    public int popSmallest() {

        int ans = this.smallIdx;

        this.arr[this.smallIdx] = true;

        this.setNextSmallestIdx(ans);

        return ans;
    }
    
    public void addBack(int num) {
        if(this.arr[num]) {
            this.arr[num] = false;
            this.smallIdx = this.smallIdx > num ? num : this.smallIdx;
        }
    }

    private void setNextSmallestIdx(int lastIdx) {
        for(int i = lastIdx; i < 1_001; ++i) {
            if(!this.arr[i]) {
                this.smallIdx = i;
                break;
            }
        }
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */

File: LeetCode/Medium/2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.kt
class SmallestInfiniteSet2() {

    val queue = PriorityQueue<Int>()
    var cur = 1

    fun popSmallest(): Int {

        if(queue.size > 0 && cur > queue.peek()) {
            return queue.poll()
        } else {
            val ans = cur
            cur++
            return ans
        }
        

    }

    fun addBack(num: Int) {
        if(num < cur && !queue.contains(num)) queue.add(num)
    }

}

class SmallestInfiniteSet() {

    val arr = BooleanArray(1_001)
    var smallIdx = 1

    fun popSmallest(): Int {

        val pop = smallIdx

        arr[smallIdx] = true

        setNextSmallIdx(pop)

        return pop
    }

    fun addBack(num: Int) {
        if(arr[num]) {
            arr[num] = false
            smallIdx = if(smallIdx > num) {num} else {smallIdx}
        }
    }

    fun setNextSmallIdx(lastIdx: Int) {
        for(i in lastIdx .. 1_000) {
            if(!arr[i]) {
                smallIdx = i
                break
            }
        }
    }

}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */